ttnn/cpp/ttnn/operations/data_movement/reshape_view/reshape.cpp

```diff
@@ -0,0 +1,168 @@
+            }
+        } else if (tensor_shape.rank() >= 2 and shape.rank() >= 2) {
+            // Handle the case when the tensor is not contiguous but the last two dimensions are the same and so reshape
+            // is possible
+            if (tensor_shape[-1] == shape[-1] and tensor_shape[-2] == shape[-2] and
+                tensor_shape.with_tile_padding()[-1] == shape.with_tile_padding()[-1] and
+                tensor_shape.with_tile_padding()[-2] == shape.with_tile_padding()[-2]) {
+                reshaped_rm_tensor = rm_tensor.reshape(shape.value);
+            }
+        } else {
+            reshaped_rm_tensor = host_reshape(tensor, shape);
+        }
+
+    }
+    // Can'd do untilize on device due to inner dim size
+    else {
+        reshaped_rm_tensor = host_reshape(tensor, shape);
+    }
+
+    if (((shape[-1] * tensor.element_size()) % sizeof(uint32_t) == 0) and reshaped_rm_tensor.layout() != layout) {
+        return ttnn::to_layout(reshaped_rm_tensor, layout, std::nullopt, std::nullopt, (Device *)nullptr);
+    }
+    else {
+        return reshaped_rm_tensor;
+    }
+
+}
+
+ttnn::Shape get_shape_from_vector(const ttnn::Tensor& tensor, const std::vector<int32_t> & shape) {
+    std::int64_t new_volume = 1;
+    std::int64_t index_of_negative_1 = -1;
+
+    for (auto index = 0; index < shape.size(); ++index) {
+        if (shape[index] == -1) {
+            if (index_of_negative_1 != -1) {
+                TT_THROW("Shape cannot have more than 1 elements that is set to -1!");
+            }
+            index_of_negative_1 = index;
+        }
+        new_volume *= shape[index];
+    }
+
+    std::vector<std::uint32_t> new_shape(shape.size());
+    std::copy(shape.begin(), shape.end(), new_shape.begin());
+    if (new_volume < 0) {
+        const auto volume = tensor.get_shape().with_tile_padding().volume();
+        new_shape[index_of_negative_1] = volume / (-new_volume);
+    }
+    return ttnn::Shape(new_shape);
+}
+
+
+}
+
+
+ttnn::Tensor ReshapeViewOperation::invoke(
+    const ttnn::Tensor& tensor,
+    const ttnn::Shape& shape
+    ) {
+
+    auto layout = tensor.get_layout();
+    auto tensor_shape = tensor.get_shape();
+    if (tensor_shape == shape) {
+        return tensor;
+    }
+    if (!(ttnn::has_storage_type_of(tensor, ttnn::StorageType::DEVICE))) {
+        return tensor.reshape(shape.value);
+
+    }
+    if (layout == ttnn::Layout::TILE and
+        ((shape.with_tile_padding()[-2] % ttnn::TILE_SIZE == 0) and (shape.with_tile_padding()[-1] % ttnn::TILE_SIZE == 0)) and
+        (tensor_shape.with_tile_padding()[-1] == shape.with_tile_padding()[-1])
+        ) {
+            return tensor.reshape(shape.value);
+    } else {
+        return detail::row_major_reshape(tensor, shape);
+    }
+
+
+
+
+}
+
+
+ttnn::Tensor ReshapeViewOperation::invoke(
+    const ttnn::Tensor& tensor,
+    const std::vector<int32_t> & shape_vector
+    ) {
+
+    auto shape = detail::get_shape_from_vector(tensor, shape_vector);
```

Comment:
shape_vector is a logical shape. it is supposed to be like that and it makes no sense creating full shape with padding from tensor, it is error prone

Reply:
this is to handle negative values for reshape. its the same code that is in core.hpp 
without this we won't be able to handle negative values.

Reply:
Oh, learned something new. Thank you!
Can you add a test with a negative value?

Reply:
there are two tests already for negative values, I copied what's in core.hpp (and will delete the old code in core.hpp) to avoid regression on the negative value tests. 
https://github.com/tenstorrent/tt-metal/blob/tarafdarTT/reshape_fix/tests/ttnn/unit_tests/test_reshape.py#L241 
https://github.com/tenstorrent/tt-metal/blob/tarafdarTT/reshape_fix/tests/ttnn/unit_tests/test_reshape.py#L176